reference/test/eigensolver/lobpcg_kernels.cpp

```diff
@@ -0,0 +1,171 @@
+        small_a_copy = gko::clone(this->small_a_cmplx);
+        small_b_copy = gko::clone(this->small_b_cmplx);
+        // LAPACK expects column-major, so transpose the matrices
+        auto small_a_t =
+            gko::share(gko::as<Mtx>(this->small_a_cmplx->transpose()));
+        auto small_b_t =
+            gko::share(gko::as<Mtx>(this->small_b_cmplx->transpose()));
+        this->small_a_cmplx = small_a_t;
+        this->small_b_cmplx = small_b_t;
+
+        small_a = this->small_a_cmplx;
+        small_b = this->small_b_cmplx;
+    } else {
+        small_a_copy = gko::clone(this->small_a_r);
+        small_b_copy = gko::clone(this->small_b_r);
+        small_a = this->small_a_r;
+        small_b = this->small_b_r;
+    }
+
+    if (std::is_same_v<gko::remove_complex<T>, gko::half>) {
+        EXPECT_THROW(
+            gko::kernels::reference::lobpcg::symm_generalized_eig(
+                this->exec, gko::kernels::lobpcg::workspace_mode::allocate,
+                small_a.get(), small_b.get(), &(this->small_e_vals), &work),
+            gko::NotImplemented);
+    } else {
+        gko::kernels::reference::lobpcg::symm_generalized_eig(
+            this->exec, gko::kernels::lobpcg::workspace_mode::allocate,
+            small_a.get(), small_b.get(), &(this->small_e_vals), &work);
+    }
+
+    if (std::is_same_v<gko::remove_complex<T>, gko::half>) return;
+
+    // On exit, the eigenvectors will be stored in the A
+    // matrix. We create submatrices for the vectors
+    // to check that A * x = lambda * B * x for each vector.
+    for (gko::size_type i = 0; i < this->small_e_vals.get_size(); i++) {
+        auto evec = gko::share(Mtx::create(
+            this->exec, gko::dim<2>{this->small_e_vals.get_size(), 1},
+            Ary::view(
+                this->exec, this->small_e_vals.get_size(),
+                small_a->get_values() + i * this->small_e_vals.get_size()),
+            1));
+
+        auto lambda_r = gko::share(Mtx_r::create(
+            this->exec, gko::dim<2>{1, 1},
+            Ary_r::view(this->exec, 1, this->small_e_vals.get_data() + i), 1));
+        std::shared_ptr<Mtx> lambda;
+        if constexpr (gko::is_complex_s<T>::value) {
+            lambda = lambda_r->make_complex();
+        } else {
+            lambda = lambda_r;
+        }
+        // A*x = lambda * B * x;
+        auto a_x = Mtx::create(this->exec,
+                               gko::dim<2>{this->small_e_vals.get_size(), 1});
+        auto lambda_b_x = Mtx::create(
+            this->exec, gko::dim<2>{this->small_e_vals.get_size(), 1});
+        lambda_b_x->fill(gko::zero<T>());
+        auto one = gko::initialize<Mtx>({gko::one<T>()}, this->exec);
+        // a_x = A * x
+        small_a_copy->apply(evec, a_x);
+        // lambda_b_x = lambda * B * x
+        small_b_copy->apply(lambda, evec, one, lambda_b_x);
+
+        const double tol =
+            100 * std::numeric_limits<gko::remove_complex<T>>::epsilon();
```

Comment:
we have r<T>::value for the purpose.

Reply:
Thanks for pointing this out -- I hadn't even noticed it (obviously, or I would've used it :) ). I changed it in 0cbb2e9c1d35cdbedc258180807ec8f3c3509162